### DISCUSSION TITLE
Phase 1

### DISCUSSION THREAD

**Discussion by Sddilora** [Ubn-Jr/hirob-be-core#13] [General]
Discussion for phase 1

**Comment by johsquaree** [comment]
Hello, good work to everyone,

I noticed the commands specified for our robot in the task list, including right, left, forward, backward, clockwise, and counterclockwise commands. I'm not sure if we should use an expression like "turn left clockwise." Can you help me with this?

**Reply by Sddilora** [reply]
According to the decision we made in our last meeting, in the first phase of the project, the robot will be a tank-like base that can only move forward and backward, turn right and left in place. The body design will be completed in the second phase. So, there is no direct movement to the right or left, but giving a command to turn right once and then move one step forward will move the robot to the right by one unit.
When the robot's commands are finished, the total distance it covers is not crucial. What's important is the sequence of steps it takes. Therefore, we will receive the commands with a message similar to "forward, forward, forward, backward, backward, clockwise, forward, counterclockwise, forward, okay".

**Comment by johsquaree** [comment]
Hello everyone, I'm currently writing tests for the create_client function, but I'm having some difficulties understanding certain aspects. I am testing the function, but all the data from the broker I connect to comes from another source, and elements like ID, name, and password are already definite and unchangeable.

I thought I could test these incoming data only when they are nil. I also found the mock package online, and it seems like it creates a clone client using it, returning them for testing, but I'm not entirely clear on these differences.

To be honest, this is my first time working with testing, and I haven't fully grasped these concepts.
...
func TestCreateClient(t *testing.T) {
        // Create a mock object.
        mockClient := new(MockMQClient)

        // createClient fTest the function.
        actualClient := createClient()
        assert.NotNil(t, actualClient) // Was the client created?

        // Was the Connect method of the mock client called?
        mockClient.On("Connect").Return(mq.NewToken(), nil)

        //Use the mockClient to mimic the operations performed by the tested function.
        client = mockClient
        actualClient = createClient()

        // Check that the assert function behaves as expected.
        mockClient.AssertExpectations(t)
    }
    ...
//mock test code

...
func TestCreateClient(t *testing.T) {
	// Save the current client to check later
	oldClient := client

	// Defer a function to reset the client after the test
	defer func() {
		client = oldClient
	}()

	// Assume a reset client
	client = nil

	// Call the createClient function
	result := createClient()

	// Check that the client is not nil
	assert.NotNil(t, result)

	// Call the createClient function once again
	// This time, check that the client is not nil and is the same as the previous one
	result2 := createClient()
	assert.NotNil(t, result2)
	assert.Equal(t, result, result2)
}
...
//my test code

**Reply by furkanadiiguzel** [reply]
There might be a problem related to usage of the mock library. This part `mockClient.On("Connect").Return(mq.NewToken(), nil)` must placed before  createClient() part

**Reply by mervegonc** [reply]
I applied to chatGPT to find a solution on the issue and the suggestions are as follows;
To ensure correct use of the mock library, you need to place the mockClient.On("Connect").Return(mq.NewToken(), nil) statement before the createClient() function. To do this your test function may need to be edited like this:

go
copy code
func TestCreateClient(t *testing.T) {
     //Create a mock object.
     mockClient := new(MockMQClient)

     // Was the Connect method of the mock client called?
     mockClient.On("Connect").Return(mq.NewToken(), nil)

     // Set the mockClient to mimic the operations performed by the tested function.
     client = mockClient

     // Test the createClient function.
     actualClient := createClient()
     assert.NotNil(t, actualClient) // Was the client created?

     // Check that the assert function behaves as expected.
     mockClient.AssertExpectations(t)
}
This arrangement ensures that mockClient.On("Connect").Return(mq.NewToken(), nil) comes before the createClient() function and can help the mock library meet expectations correctly.

**Reply by Sddilora** [reply]
@johsquaree Could you please rearranging the test based on the given feedback, review and share your thoughts on whether these suggestions were suitable for the test ?
